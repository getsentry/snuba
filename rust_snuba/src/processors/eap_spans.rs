use anyhow::Context;
use chrono::DateTime;
use seq_macro::seq;
use serde::Serialize;
use std::collections::HashMap;
use uuid::Uuid;

use rust_arroyo::backends::kafka::types::KafkaPayload;

use crate::config::ProcessorConfig;
use crate::processors::spans::FromSpanMessage;
use crate::processors::utils::enforce_retention;
use crate::types::{InsertBatch, KafkaMessageMetadata};

pub fn process_message(
    payload: KafkaPayload,
    _metadata: KafkaMessageMetadata,
    config: &ProcessorConfig,
) -> anyhow::Result<InsertBatch> {
    let payload_bytes = payload.payload().context("Expected payload")?;
    let msg: FromSpanMessage = serde_json::from_slice(payload_bytes)?;

    let origin_timestamp = DateTime::from_timestamp(msg.received as i64, 0);
    let mut span: EAPSpan = msg.try_into()?;

    span.retention_days = Some(enforce_retention(span.retention_days, &config.env_config));

    InsertBatch::from_rows([span], origin_timestamp)
}

seq!(N in 0..200 {
#[derive(Debug, Default, Serialize)]
struct EAPSpan {
    // the span object for the new "events analytics platform"
    organization_id: u64,
    project_id: u64,
    service: String, //currently just project ID as a string
    trace_id: Uuid,
    span_id: u64,
    #[serde(default)]
    parent_span_id: u64,
    segment_id: u64,      //aka transaction ID
    segment_name: String, //aka transaction name
    is_segment: bool,     //aka "is transaction"
    start_timestamp: u64,
    end_timestamp: u64,
    duration_ms: u32,
    exclusive_time_ms: f64,
    retention_days: Option<u16>,
    name: String, //aka description

    sampling_factor: f64,
    sampling_weight: f64,
    sign: u8, //1 for additions, -1 for deletions - for this worker it should be 1

    #(
    #[serde(skip_serializing_if = "HashMap::is_empty")]
    attr_str_~N: HashMap<String, String>,

    #[serde(skip_serializing_if = "HashMap::is_empty")]
    attr_num_~N: HashMap<String, f64>,
    )*
}
});

fn fnv_1a(input: &[u8]) -> u32 {
    const FNV_1A_PRIME: u32 = 16777619;
    const FNV_1A_OFFSET_BASIS: u32 = 2166136261;

    let mut res = FNV_1A_OFFSET_BASIS;
    for byt in input {
        res ^= *byt as u32;
        res = res.wrapping_mul(FNV_1A_PRIME);
    }

    res
}

impl From<FromSpanMessage> for EAPSpan {
    fn from(from: FromSpanMessage) -> EAPSpan {
        let sentry_tags = from.sentry_tags.unwrap_or_default();
        let tags = from.tags.unwrap_or_default();
        let measurements = from.measurements.unwrap_or_default();

        let mut res = Self {
            organization_id: from.organization_id,
            project_id: from.project_id,
            trace_id: from.trace_id,
            service: format!("{:x}", from.project_id),
            span_id: u64::from_str_radix(&from.span_id, 16).unwrap_or_default(),
            parent_span_id: from
                .parent_span_id
                .map_or(0, |s| u64::from_str_radix(&s, 16).unwrap_or(0)),
            segment_id: from
                .segment_id
                .map_or(0, |s| u64::from_str_radix(&s, 16).unwrap_or(0)),
            segment_name: sentry_tags.get("transaction").cloned().unwrap_or_default(),
            is_segment: from.is_segment,
            start_timestamp: (from.start_timestamp_precise * 1e6) as u64,
            end_timestamp: (from.end_timestamp_precise * 1e6) as u64,
            duration_ms: from.duration_ms,
            exclusive_time_ms: from.exclusive_time_ms,
            retention_days: from.retention_days,
            name: from.description.unwrap_or_default(),

            sampling_weight: 1.,
            sampling_factor: 1.,
            sign: 1,

            ..Default::default()
        };

        {
            seq!(N in 0..200 {
            let mut attr_str_buckets = [
                #(
                &mut res.attr_str_~N,
                )*
            ];
            });

            sentry_tags.iter().chain(tags.iter()).for_each(|(k, v)| {
                attr_str_buckets[(fnv_1a(k.as_bytes()) as usize) % attr_str_buckets.len()]
                    .insert(k.clone(), v.clone());
            });
        }

        {
            seq!(N in 0..200 {
            let mut attr_num_buckets = [
                #(
                &mut res.attr_num_~N,
                )*
            ];
                });

            measurements.iter().for_each(|(k, v)| {
                attr_num_buckets[(fnv_1a(k.as_bytes()) as usize) % attr_num_buckets.len()]
                    .insert(k.clone(), v.value);
            });
        }

        res
    }
}

#[cfg(test)]
mod tests {
    use std::time::SystemTime;

    use super::*;

    const SPAN_KAFKA_MESSAGE: &str = r#"
{
    "description": "/api/0/relays/projectconfigs/",
    "duration_ms": 152,
    "event_id": "d826225de75d42d6b2f01b957d51f18f",
    "exclusive_time_ms": 0.228,
    "is_segment": true,
    "data": {
        "sentry.environment": "development",
        "sentry.release": "backend@24.7.0.dev0+c45b49caed1e5fcbf70097ab3f434b487c359b6b",
        "thread.name": "uWSGIWorker1Core0",
        "thread.id": "8522009600",
        "sentry.segment.name": "/api/0/relays/projectconfigs/",
        "sentry.sdk.name": "sentry.python.django",
        "sentry.sdk.version": "2.7.0"
    },
    "measurements": {
        "num_of_spans": {
            "value": 50.0
        }
    },
    "organization_id": 1,
    "origin": "auto.http.django",
    "project_id": 1,
    "received": 1721319572.877828,
    "retention_days": 90,
    "segment_id": "8873a98879faf06d",
    "sentry_tags": {
        "category": "http",
        "environment": "development",
        "op": "http.server",
        "platform": "python",
        "release": "backend@24.7.0.dev0+c45b49caed1e5fcbf70097ab3f434b487c359b6b",
        "sdk.name": "sentry.python.django",
        "sdk.version": "2.7.0",
        "status": "ok",
        "status_code": "200",
        "thread.id": "8522009600",
        "thread.name": "uWSGIWorker1Core0",
        "trace.status": "ok",
        "transaction": "/api/0/relays/projectconfigs/",
        "transaction.method": "POST",
        "transaction.op": "http.server",
        "user": "ip:127.0.0.1"
    },
    "span_id": "8873a98879faf06d",
    "tags": {
        "http.status_code": "200",
        "relay_endpoint_version": "3",
        "relay_id": "88888888-4444-4444-8444-cccccccccccc",
        "relay_no_cache": "False",
        "relay_protocol_version": "3",
        "relay_use_post_or_schedule": "True",
        "relay_use_post_or_schedule_rejected": "version",
        "server_name": "D23CXQ4GK2.local",
        "spans_over_limit": "False"
    },
    "trace_id": "d099bf9ad5a143cf8f83a98081d0ed3b",
    "start_timestamp_ms": 1721319572616,
    "start_timestamp_precise": 1721319572.616648,
    "end_timestamp_precise": 1721319572.768806
}
    "#;

    #[test]
    fn test_half_md5() {
        //select halfMD5('test') == 688887797400064883
        assert_eq!(fnv_1a("test".as_bytes()), 2949673445)
    }

    #[test]
    fn test_valid_span() {
        let payload = KafkaPayload::new(None, None, Some(SPAN_KAFKA_MESSAGE.as_bytes().to_vec()));
        let meta = KafkaMessageMetadata {
            partition: 0,
            offset: 1,
            timestamp: DateTime::from(SystemTime::now()),
        };
        process_message(payload, meta, &ProcessorConfig::default())
            .expect("The message should be processed");
    }

    #[test]
    fn test_serialization() {
        let msg: FromSpanMessage = serde_json::from_slice(SPAN_KAFKA_MESSAGE.as_bytes()).unwrap();
        let span: EAPSpan = msg.try_into().unwrap();
        insta::with_settings!({sort_maps => true}, {
            insta::assert_json_snapshot!(span)
        });
    }
}
