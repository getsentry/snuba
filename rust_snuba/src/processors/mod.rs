mod functions;
mod profiles;
mod querylog;
mod spans;
mod utils;

use crate::types::{KafkaMessageMetadata, RowData};
use rust_arroyo::backends::kafka::types::KafkaPayload;

type ProcessingFunction = fn(KafkaPayload, KafkaMessageMetadata) -> anyhow::Result<RowData>;

pub fn get_processing_function(name: &str) -> Option<ProcessingFunction> {
    match name {
        "FunctionsMessageProcessor" => Some(functions::process_message),
        "ProfilesMessageProcessor" => Some(profiles::process_message),
        "QuerylogProcessor" => Some(querylog::process_message),
        "SpansMessageProcessor" => Some(spans::process_message),
        _ => None,
    }
}
