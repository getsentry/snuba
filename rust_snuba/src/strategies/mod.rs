pub mod clickhouse;
pub mod commit_log;
pub mod python;
pub mod validate_schema;
