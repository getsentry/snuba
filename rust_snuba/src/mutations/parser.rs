use std::collections::BTreeMap;

use rust_arroyo::backends::kafka::types::KafkaPayload;
use rust_arroyo::counter;
use rust_arroyo::processing::strategies::run_task_in_threads::{
    RunTaskError, RunTaskFunc, TaskRunner,
};
use rust_arroyo::types::Message;

use schemars::JsonSchema;
use sentry::Hub;
use sentry::SentryFutureExt;
use serde::Deserialize;

use crate::arroyo_utils::invalid_message_err;
use crate::processors::eap_spans::{FromPrimaryKey, PrimaryKey};

#[derive(Debug, Default, Deserialize, JsonSchema)]
pub(crate) struct Update {
    // the update clause
    pub attr_str: BTreeMap<String, String>,
    pub attr_num: BTreeMap<String, f64>,
}

impl Update {
    pub fn merge(&mut self, other: Update) {
        self.attr_str.extend(other.attr_str);
        self.attr_num.extend(other.attr_num);
    }
}

#[derive(Debug, Default, Deserialize, JsonSchema)]
pub(crate) struct MutationMessage {
    // primary key, the mutation only applies on the rows that match this filter
    pub filter: FromPrimaryKey,

    // the span attributes to update
    pub update: Update,
}

#[derive(Default)]
pub(crate) struct MutationBatch(pub BTreeMap<PrimaryKey, Update>);

#[derive(Clone)]
pub(crate) struct MutationParser;

impl MutationParser {
    async fn process_message(
        self,
        message: Message<KafkaPayload>,
    ) -> Result<Message<MutationMessage>, RunTaskError<anyhow::Error>> {
        let maybe_err = RunTaskError::InvalidMessage(
            invalid_message_err(&message).map_err(RunTaskError::Other)?,
        );

        message
            .try_map(|payload| {
                let payload = payload
                    .payload()
                    .ok_or(anyhow::anyhow!("no payload in message"))?;
                let parsed: MutationMessage = serde_json::from_slice(payload)?;
                Ok(parsed)
            })
            .map_err(|error: anyhow::Error| {
                let error: &dyn std::error::Error = error.as_ref();
                counter!("invalid_mutation");
                tracing::error!(error, "failed processing mutation");
                maybe_err
            })
    }
}

impl TaskRunner<KafkaPayload, MutationMessage, anyhow::Error> for MutationParser {
    fn get_task(
        &self,
        message: Message<KafkaPayload>,
    ) -> RunTaskFunc<MutationMessage, anyhow::Error> {
        Box::pin(
            self.clone()
                .process_message(message)
                .bind_hub(Hub::new_from_top(Hub::current())),
        )
    }
}
