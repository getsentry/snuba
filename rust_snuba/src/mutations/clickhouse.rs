use rust_arroyo::counter;
use rust_arroyo::processing::strategies::run_task_in_threads::{
    RunTaskError, RunTaskFunc, TaskRunner,
};
use rust_arroyo::types::Message;
use serde::Serialize;
use std::time::Duration;

use reqwest::header::{HeaderMap, HeaderValue, ACCEPT_ENCODING, CONNECTION};
use reqwest::{Client, ClientBuilder};
use uuid::Uuid;

use crate::mutations::parser::MutationBatch;
use crate::processors::eap_spans::{AttributeMap, PrimaryKey, ATTRS_SHARD_FACTOR};
#[derive(Clone)]
pub struct ClickhouseWriter {
    url: String,
    table: String,
    client: Client,
}

impl ClickhouseWriter {
    pub fn new(
        hostname: &str,
        http_port: u16,
        table: &str,
        database: &str,
        clickhouse_user: &str,
        clickhouse_password: &str,
        batch_write_timeout: Option<Duration>,
    ) -> Self {
        let mut headers = HeaderMap::with_capacity(5);
        headers.insert(CONNECTION, HeaderValue::from_static("keep-alive"));
        headers.insert(ACCEPT_ENCODING, HeaderValue::from_static("gzip,deflate"));
        headers.insert(
            "X-Clickhouse-User",
            HeaderValue::from_str(clickhouse_user).unwrap(),
        );
        headers.insert(
            "X-ClickHouse-Key",
            HeaderValue::from_str(clickhouse_password).unwrap(),
        );
        headers.insert(
            "X-ClickHouse-Database",
            HeaderValue::from_str(database).unwrap(),
        );

        let url = format!("http://{hostname}:{http_port}");

        let mut client_builder = ClientBuilder::new().default_headers(headers);

        if let Some(timeout) = batch_write_timeout {
            client_builder = client_builder.timeout(timeout);
        }

        Self {
            url,
            table: table.to_owned(),
            client: client_builder.build().unwrap(),
        }
    }

    async fn process_message(&self, message: &Message<MutationBatch>) -> anyhow::Result<()> {
        let queries = format_query(&self.table, message.payload());
        let session_id = Uuid::new_v4().to_string();

        for query in queries {
            let response = self
                .client
                .post(&self.url)
                .query(&[("session_id", &session_id)])
                .body(query)
                .send()
                .await?;

            if !response.status().is_success() {
                let status = response.status();
                let body = response.text().await;
                anyhow::bail!(
                    "bad response while inserting mutation, status: {}, response body: {:?}",
                    status,
                    body
                );
            }
        }

        Ok(())
    }
}

impl TaskRunner<MutationBatch, (), anyhow::Error> for ClickhouseWriter {
    fn get_task(&self, message: Message<MutationBatch>) -> RunTaskFunc<(), anyhow::Error> {
        let slf = self.clone();

        Box::pin(async move {
            slf.process_message(&message)
                .await
                .map_err(RunTaskError::Other)?;
            message.try_map(|_| Ok(()))
        })
    }
}

fn format_query(table: &str, batch: &MutationBatch) -> Vec<Vec<u8>> {
    let mut attr_columns = String::new();
    // attr_combined_columns is intentionally ordered the same as the clickhouse schema.
    // INSERT INTO .. SELECT FROM .. matches up columns by position only, and ignores names.
    // subqueries don't have this property.
    let mut attr_combined_columns = String::new();
    for i in 0..ATTRS_SHARD_FACTOR {
        attr_columns.push_str(&format!(",attr_str_{i} Map(String, String)"));

        attr_combined_columns.push_str(&format!(
            ",if(sign = 1, mapUpdate(old_data.attr_str_{i}, new_data.attr_str_{i}), old_data.attr_str_{i}) AS attr_str_{i}"
        ));
    }

    for i in 0..ATTRS_SHARD_FACTOR {
        attr_columns.push_str(&format!(",attr_num_{i} Map(String, Float64)"));
        attr_combined_columns.push_str(&format!(
            ",if(sign = 1, mapUpdate(old_data.attr_num_{i}, new_data.attr_num_{i}), old_data.attr_num_{i}) AS attr_num_{i}"
        ));
    }

    let input_schema = format!("organization_id UInt64, _sort_timestamp DateTime, trace_id UUID, span_id UInt64 {attr_columns}");
    let create_tmp_table = format!("CREATE TEMPORARY TABLE new_data ({input_schema})").into_bytes();
    let mut insert_tmp_table = "INSERT INTO new_data FORMAT JSONEachRow\n"
        .to_owned()
        .into_bytes();

    counter!("eap_mutations.mutation_rows", batch.0.len() as u64);
    for (filter, update) in &batch.0 {
        let mut attributes = AttributeMap::default();
        for (k, v) in &update.attr_str {
            attributes.insert_str(k.clone(), v.clone());
        }

        for (k, v) in &update.attr_num {
            attributes.insert_num(k.clone(), *v);
        }

        let row = MutationRow {
            filter: filter.clone(),
            attributes,
        };

        serde_json::to_writer(&mut insert_tmp_table, &row).unwrap();
        insert_tmp_table.push(b'\n');
    }

    let main_insert = format!(
        "
        INSERT INTO {table}
        SELECT old_data.* EXCEPT ('sign|attr_.*'), arrayJoin([-1, 1]) as sign {attr_combined_columns}
        FROM {table} old_data
        GLOBAL JOIN new_data
        ON old_data.organization_id = new_data.organization_id
            and old_data.trace_id = new_data.trace_id
            and old_data.span_id = new_data.span_id
            and old_data._sort_timestamp = new_data._sort_timestamp
        PREWHERE
        (old_data.organization_id,
            old_data._sort_timestamp,
            old_data.trace_id,
            old_data.span_id,
            ) GLOBAL IN (SELECT organization_id, _sort_timestamp, trace_id, span_id FROM new_data)
        WHERE
        old_data.sign = 1
        "
    )
    .into_bytes();

    let drop_tmp_table = "DROP TABLE IF EXISTS new_data".to_owned().into_bytes();

    vec![
        create_tmp_table,
        insert_tmp_table,
        main_insert,
        drop_tmp_table,
    ]
}

#[derive(Serialize, Default)]
struct MutationRow {
    #[serde(flatten)]
    attributes: AttributeMap,

    #[serde(flatten)]
    filter: PrimaryKey,
}

#[cfg(test)]
mod tests {
    use std::env;
    use uuid::Uuid;

    use crate::mutations::parser::Update;

    use super::*;

    struct ClickhouseTestClient {
        url: String,
        table: String,
        client: Client,
    }

    impl ClickhouseTestClient {
        pub async fn new(table: String) -> anyhow::Result<Self> {
            let hostname = env::var("CLICKHOUSE_HOST").unwrap_or_else(|_| "127.0.0.1".to_owned());
            let http_port = 8123;
            let url = format!("http://{hostname}:{http_port}");
            let uuid = Uuid::new_v4().to_string();
            let table_id = uuid.split('-').next().unwrap();

            let client = reqwest::Client::new();
            let test_table = format!("{table}_{table_id}_local\n");

            let body =
                format!("CREATE TABLE IF NOT EXISTS {test_table} AS {table}_local\n").into_bytes();

            // use client to create a new table locally
            client.post(url.clone()).body(body).send().await?;

            Ok(Self {
                url,
                table: test_table,
                client,
            })
        }

        pub async fn run_mutation(&self, queries: Vec<Vec<u8>>) -> anyhow::Result<()> {
            let session_id = Uuid::new_v4().to_string();

            for query in queries {
                self.client
                    .post(&self.url)
                    .query(&[("session_id", &session_id)])
                    .body(query)
                    .send()
                    .await?;
            }
            Ok(())
        }

        pub async fn select_final(&self) -> anyhow::Result<String> {
            let table = &self.table;
            let final_query = format!("SELECT * FROM {table} FINAL\n").into_bytes();

            let response = self.client.post(&self.url).body(final_query).send().await?;

            let update = response.text().await?;

            Ok(update)
        }

        pub async fn drop_table(&self) -> anyhow::Result<()> {
            let table = &self.table;
            let drop = format!("DROP TABLE {table}\n").into_bytes();

            self.client.post(&self.url).body(drop).send().await?;

            Ok(())
        }
    }

    #[tokio::test]
    async fn test_simple_mutation() {
        let mut batch = MutationBatch::default();
        let mut update = Update::default();

        let organization_id = 69;
        let _sort_timestamp = 1727466947;
        let trace_id = Uuid::parse_str("deadbeef-dead-beef-dead-beefdeadbeef").unwrap();
        let span_id = 16045690984833335023;

        let primary_key = PrimaryKey {
            organization_id,
            _sort_timestamp,
            trace_id,
            span_id,
        };
        update.attr_str.insert("a".to_string(), "b".to_string());

        // build the mutation batch
        batch.0.insert(primary_key.clone(), update);

        let test_client = ClickhouseTestClient::new("eap_spans".to_string())
            .await
            .unwrap();

        let test_table = &test_client.table;

        let insert =
        format!("INSERT INTO {test_table} (organization_id, _sort_timestamp, trace_id, span_id, sign, retention_days) VALUES ({organization_id}, {_sort_timestamp}, \'{trace_id}\', {span_id}, 1, 90)\n").into_bytes();

        let _ = test_client
            .client
            .post(&test_client.url)
            .body(insert)
            .send()
            .await;

        let all_queries = format_query(test_table, &batch);
        let _ = test_client.run_mutation(all_queries).await;

        // merge data at query time for up-to-date results
        let mutation = test_client.select_final().await;
        assert!(mutation.unwrap().contains("{'a':'b'}"));

        // clean up the temporary table at the end of test
        let _ = test_client.drop_table().await;
    }

    #[test]
    fn format_query_snap() {
        let mut batch = MutationBatch::default();
        let mut update = Update::default();
        let primary_key = PrimaryKey {
            organization_id: 69,
            _sort_timestamp: 1727466947,
            trace_id: Uuid::parse_str("deadbeef-dead-beef-dead-beefdeadbeef").unwrap(),
            span_id: 16045690984833335023,
        };
        update.attr_str.insert("a".to_string(), "b".to_string());

        batch.0.insert(primary_key.clone(), update);

        let mut snapshot = String::new();
        for query in format_query("eap_spans_local", &batch) {
            snapshot.push_str(std::str::from_utf8(&query).unwrap());
            snapshot.push_str(";\n");
        }

        insta::assert_snapshot!(snapshot);
    }
}
