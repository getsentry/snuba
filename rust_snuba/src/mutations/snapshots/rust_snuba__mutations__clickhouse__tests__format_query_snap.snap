---
source: src/mutations/clickhouse.rs
expression: snapshot
---
CREATE TEMPORARY TABLE new_data (organization_id UInt64, _sort_timestamp DateTime, trace_id UUID, span_id UInt64 ,attr_str_0 Map(String, String),attr_str_1 Map(String, String),attr_str_2 Map(String, String),attr_str_3 Map(String, String),attr_str_4 Map(String, String),attr_str_5 Map(String, String),attr_str_6 Map(String, String),attr_str_7 Map(String, String),attr_str_8 Map(String, String),attr_str_9 Map(String, String),attr_str_10 Map(String, String),attr_str_11 Map(String, String),attr_str_12 Map(String, String),attr_str_13 Map(String, String),attr_str_14 Map(String, String),attr_str_15 Map(String, String),attr_str_16 Map(String, String),attr_str_17 Map(String, String),attr_str_18 Map(String, String),attr_str_19 Map(String, String),attr_num_0 Map(String, Float64),attr_num_1 Map(String, Float64),attr_num_2 Map(String, Float64),attr_num_3 Map(String, Float64),attr_num_4 Map(String, Float64),attr_num_5 Map(String, Float64),attr_num_6 Map(String, Float64),attr_num_7 Map(String, Float64),attr_num_8 Map(String, Float64),attr_num_9 Map(String, Float64),attr_num_10 Map(String, Float64),attr_num_11 Map(String, Float64),attr_num_12 Map(String, Float64),attr_num_13 Map(String, Float64),attr_num_14 Map(String, Float64),attr_num_15 Map(String, Float64),attr_num_16 Map(String, Float64),attr_num_17 Map(String, Float64),attr_num_18 Map(String, Float64),attr_num_19 Map(String, Float64));
INSERT INTO new_data FORMAT JSONEachRow
{"attr_str_0":{"a":"b"},"organization_id":69,"_sort_timestamp":1727466947,"trace_id":"deadbeef-dead-beef-dead-beefdeadbeef","span_id":16045690984833335023}
;

        INSERT INTO eap_spans_local
        SELECT old_data.* EXCEPT ('sign|attr_.*'), arrayJoin([-1, 1]) as sign ,if(sign = 1, mapUpdate(old_data.attr_str_0, new_data.attr_str_0), old_data.attr_str_0) AS attr_str_0,if(sign = 1, mapUpdate(old_data.attr_str_1, new_data.attr_str_1), old_data.attr_str_1) AS attr_str_1,if(sign = 1, mapUpdate(old_data.attr_str_2, new_data.attr_str_2), old_data.attr_str_2) AS attr_str_2,if(sign = 1, mapUpdate(old_data.attr_str_3, new_data.attr_str_3), old_data.attr_str_3) AS attr_str_3,if(sign = 1, mapUpdate(old_data.attr_str_4, new_data.attr_str_4), old_data.attr_str_4) AS attr_str_4,if(sign = 1, mapUpdate(old_data.attr_str_5, new_data.attr_str_5), old_data.attr_str_5) AS attr_str_5,if(sign = 1, mapUpdate(old_data.attr_str_6, new_data.attr_str_6), old_data.attr_str_6) AS attr_str_6,if(sign = 1, mapUpdate(old_data.attr_str_7, new_data.attr_str_7), old_data.attr_str_7) AS attr_str_7,if(sign = 1, mapUpdate(old_data.attr_str_8, new_data.attr_str_8), old_data.attr_str_8) AS attr_str_8,if(sign = 1, mapUpdate(old_data.attr_str_9, new_data.attr_str_9), old_data.attr_str_9) AS attr_str_9,if(sign = 1, mapUpdate(old_data.attr_str_10, new_data.attr_str_10), old_data.attr_str_10) AS attr_str_10,if(sign = 1, mapUpdate(old_data.attr_str_11, new_data.attr_str_11), old_data.attr_str_11) AS attr_str_11,if(sign = 1, mapUpdate(old_data.attr_str_12, new_data.attr_str_12), old_data.attr_str_12) AS attr_str_12,if(sign = 1, mapUpdate(old_data.attr_str_13, new_data.attr_str_13), old_data.attr_str_13) AS attr_str_13,if(sign = 1, mapUpdate(old_data.attr_str_14, new_data.attr_str_14), old_data.attr_str_14) AS attr_str_14,if(sign = 1, mapUpdate(old_data.attr_str_15, new_data.attr_str_15), old_data.attr_str_15) AS attr_str_15,if(sign = 1, mapUpdate(old_data.attr_str_16, new_data.attr_str_16), old_data.attr_str_16) AS attr_str_16,if(sign = 1, mapUpdate(old_data.attr_str_17, new_data.attr_str_17), old_data.attr_str_17) AS attr_str_17,if(sign = 1, mapUpdate(old_data.attr_str_18, new_data.attr_str_18), old_data.attr_str_18) AS attr_str_18,if(sign = 1, mapUpdate(old_data.attr_str_19, new_data.attr_str_19), old_data.attr_str_19) AS attr_str_19,if(sign = 1, mapUpdate(old_data.attr_num_0, new_data.attr_num_0), old_data.attr_num_0) AS attr_num_0,if(sign = 1, mapUpdate(old_data.attr_num_1, new_data.attr_num_1), old_data.attr_num_1) AS attr_num_1,if(sign = 1, mapUpdate(old_data.attr_num_2, new_data.attr_num_2), old_data.attr_num_2) AS attr_num_2,if(sign = 1, mapUpdate(old_data.attr_num_3, new_data.attr_num_3), old_data.attr_num_3) AS attr_num_3,if(sign = 1, mapUpdate(old_data.attr_num_4, new_data.attr_num_4), old_data.attr_num_4) AS attr_num_4,if(sign = 1, mapUpdate(old_data.attr_num_5, new_data.attr_num_5), old_data.attr_num_5) AS attr_num_5,if(sign = 1, mapUpdate(old_data.attr_num_6, new_data.attr_num_6), old_data.attr_num_6) AS attr_num_6,if(sign = 1, mapUpdate(old_data.attr_num_7, new_data.attr_num_7), old_data.attr_num_7) AS attr_num_7,if(sign = 1, mapUpdate(old_data.attr_num_8, new_data.attr_num_8), old_data.attr_num_8) AS attr_num_8,if(sign = 1, mapUpdate(old_data.attr_num_9, new_data.attr_num_9), old_data.attr_num_9) AS attr_num_9,if(sign = 1, mapUpdate(old_data.attr_num_10, new_data.attr_num_10), old_data.attr_num_10) AS attr_num_10,if(sign = 1, mapUpdate(old_data.attr_num_11, new_data.attr_num_11), old_data.attr_num_11) AS attr_num_11,if(sign = 1, mapUpdate(old_data.attr_num_12, new_data.attr_num_12), old_data.attr_num_12) AS attr_num_12,if(sign = 1, mapUpdate(old_data.attr_num_13, new_data.attr_num_13), old_data.attr_num_13) AS attr_num_13,if(sign = 1, mapUpdate(old_data.attr_num_14, new_data.attr_num_14), old_data.attr_num_14) AS attr_num_14,if(sign = 1, mapUpdate(old_data.attr_num_15, new_data.attr_num_15), old_data.attr_num_15) AS attr_num_15,if(sign = 1, mapUpdate(old_data.attr_num_16, new_data.attr_num_16), old_data.attr_num_16) AS attr_num_16,if(sign = 1, mapUpdate(old_data.attr_num_17, new_data.attr_num_17), old_data.attr_num_17) AS attr_num_17,if(sign = 1, mapUpdate(old_data.attr_num_18, new_data.attr_num_18), old_data.attr_num_18) AS attr_num_18,if(sign = 1, mapUpdate(old_data.attr_num_19, new_data.attr_num_19), old_data.attr_num_19) AS attr_num_19
        FROM eap_spans_local old_data
        GLOBAL JOIN new_data
        ON old_data.organization_id = new_data.organization_id
            and old_data.trace_id = new_data.trace_id
            and old_data.span_id = new_data.span_id
            and old_data._sort_timestamp = new_data._sort_timestamp
        PREWHERE
        (old_data.organization_id,
            old_data._sort_timestamp,
            old_data.trace_id,
            old_data.span_id,
            ) GLOBAL IN (SELECT organization_id, _sort_timestamp, trace_id, span_id FROM new_data)
        WHERE
        old_data.sign = 1
        ;
DROP TABLE IF EXISTS new_data;
